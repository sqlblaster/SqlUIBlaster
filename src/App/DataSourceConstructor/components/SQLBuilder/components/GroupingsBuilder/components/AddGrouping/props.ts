import { actionCreators, PickActionCreators } from 'src/store/action-creators';
import { PickStates } from 'src/store/models/State';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';

export type AddGroupingOwnProps = PopoverManagerProps;

export type AddGroupingStateProps = PickStates<'groupings'>;

export type AddGroupingDispatchProps = PickActionCreators<'addGrouping'>;

export const AGMapDispatchToProps: AddGroupingDispatchProps = {
  addGrouping: actionCreators.addGrouping
};

export type AddGroupingProps = AddGroupingStateProps & AddGroupingDispatchProps;

export type AddGroupingViewProps = AddGroupingOwnProps &
  AddGroupingStateProps &
  AddGroupingDispatchProps;
