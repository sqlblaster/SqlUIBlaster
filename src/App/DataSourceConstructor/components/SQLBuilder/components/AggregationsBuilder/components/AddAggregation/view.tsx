import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AddAggregationViewProps } from './props';

export const addAggregationText = 'Add an aggregation';

// FIXME: make padding clickable too
export class AddAggregation extends React.Component<AddAggregationViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddAggregationViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      aggregations,
      addAggregation,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(aggregations && aggregations.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addAggregationText}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateAggregationPopover
            action={addAggregation}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
