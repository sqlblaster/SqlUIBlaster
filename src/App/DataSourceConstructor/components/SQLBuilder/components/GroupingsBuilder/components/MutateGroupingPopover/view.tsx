import * as React from 'react';
import { ColumnSelector } from '../../../common/ColumnSelector';
import { PopoverPaper } from '../../../common/PopoverPaper';
import { MutateGroupingPopoverViewProps } from './props';

export const MutateGroupingPopover: React.FC<
  MutateGroupingPopoverViewProps
> = props => {
  const {
    popoverPositionUpdater,
    onColumnSelected,
    canShowColumnBranch
  } = props;

  return (
    <PopoverPaper {...props}>
      <div className='mutate-grouping'>
        <ColumnSelector
          popoverPositionUpdater={popoverPositionUpdater}
          onColumnSelected={onColumnSelected}
          canShowColumnBranch={canShowColumnBranch}
        />
      </div>
    </PopoverPaper>
  );
};
