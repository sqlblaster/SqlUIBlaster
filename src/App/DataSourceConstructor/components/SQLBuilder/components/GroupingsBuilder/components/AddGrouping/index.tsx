import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/components/query-builder/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  AddGroupingDispatchProps,
  AddGroupingProps,
  AddGroupingStateProps,
  AGMapDispatchToProps
} from './props';
import { AddGrouping as View } from './view';

export const AddGrouping = connect<
  AddGroupingStateProps,
  AddGroupingDispatchProps,
  {},
  State
>(
  ({ groupings }) => ({
    groupings
  }),
  AGMapDispatchToProps
)(
  class extends React.Component<AddGroupingProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public render() {
      return (
        <AddGrouping.ViewWithPopoverManagement
          {...this.props}
        />
      );
    }
  }
);
