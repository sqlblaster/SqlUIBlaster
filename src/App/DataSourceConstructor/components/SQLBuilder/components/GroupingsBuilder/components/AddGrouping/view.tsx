import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateGroupingPopover } from '../MutateGroupingPopover';
import { AddGroupingViewProps } from './props';

export const addGroupingText = 'Add a grouping';

// FIXME: make padding clickable too
export class AddGrouping extends React.Component<AddGroupingViewProps> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddGroupingViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      groupings,
      addGrouping,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(groupings && groupings.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addGroupingText}
        </AddEntity>
        {isOpen && (// TODO: resolve this problem another way
          <MutateGroupingPopover
            action={addGrouping}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        )}
      </>
    );
  }
}
