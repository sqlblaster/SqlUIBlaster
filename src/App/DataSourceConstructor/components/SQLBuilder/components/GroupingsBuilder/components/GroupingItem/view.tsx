import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { MutateGroupingPopover } from '../MutateGroupingPopover';
import { GroupingItemViewProps } from './props';
import { groupingItemStyles } from './styles';

export const GroupingItem = withStyles(groupingItemStyles)(
  class extends React.Component<GroupingItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: GroupingItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        grouping,
        updateGrouping,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleGroupingRemoval
      } = this.props;

      const { id, column } = grouping;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
              data-testid='column-view'
            >
              <SequentialColumnView {...column} />
            </div>
            <RemoveButton
              onClick={handleGroupingRemoval(id)}
              data-testid='remove'
            />
          </div>
          { isOpen && (
            <MutateGroupingPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateGrouping}
              onClose={handleClose}
              grouping={grouping}
            />
          ) }
        </>
      );
    }
  }
);
