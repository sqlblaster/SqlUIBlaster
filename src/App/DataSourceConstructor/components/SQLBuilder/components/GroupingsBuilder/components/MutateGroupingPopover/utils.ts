import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { Grouping } from '../GroupingItem/model';

export const filterGrouping = (
  columnBranch: ColumnBranch,
  groupings: Grouping[]
): boolean => !groupings.some(({ column }) => column && column.equals(columnBranch));
