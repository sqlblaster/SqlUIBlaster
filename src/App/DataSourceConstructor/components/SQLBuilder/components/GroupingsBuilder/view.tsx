import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { AddGrouping } from './components/AddGrouping';
import { GroupingItem } from './components/GroupingItem';
import { GroupingsBuilderViewProps } from './props';
import { groupingsBuilderStyles } from './styles';

export const GroupingsBuilder = withStyles(groupingsBuilderStyles)((({
  groupings,
  classes
}) => (
  <div className={classes.root}>
    { groupings && groupings.length ? (
      <div className={classes.groupings}>
        { groupings.map((grouping) => (
          <GroupingItem key={grouping.id} grouping={grouping} />
        )) }
      </div>
    ) : null }
    <AddGrouping />
  </div>
)) as React.FC<GroupingsBuilderViewProps>);
