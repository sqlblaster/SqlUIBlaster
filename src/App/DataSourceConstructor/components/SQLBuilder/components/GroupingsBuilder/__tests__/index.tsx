import {
  fireEvent,
  getByText as getByTextOn,
  queryByText as queryByTextOn
} from 'react-testing-library';
import { regex } from 'src/components/query-builder/App/DataSourceConstructor/test-utils/utils';
import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { groupingsBuilderLabel } from '../../../labels';
import { selectAColumn } from '../../common/ColumnSelector/__tests__/utils/select-a-column';
import {
  clickAddAGrouping,
  clickAddMoreGrouping
} from '../../GroupingsBuilder/components/AddGrouping/__tests__/utils/click-add-a-grouping';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import { addGroupingText } from '../components/AddGrouping/view';

describe('Groupings Builder', () => {
  const { getByText, queryByText, getAllByText, getByTestId } = render();

  const columns = selectATable();

  const groupingsBuilderPanel = getByTextOn(
    document.body,
    regex(groupingsBuilderLabel)
  );

  const clickOnGrouping = (columnName: string) => {
    fireEvent.click(getByTextOn(groupingsBuilderPanel, regex(columnName)));
  };

  const removeAGrouping = () => {
    fireEvent.click(getAllByText(/clear/i)[0]);
  };

  const detailsColumn = columns[1];
  const createdAtColumn = columns[2];

  it('should show the add grouping placeholder', () => {
    expect(getByText(addGroupingText)).toBeInTheDocument();
  });

  it(`should add a grouping`, () => {
    clickAddAGrouping();

    fireEvent.click(getByText(detailsColumn.name));

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should hide the add grouping placeholder', () => {
    expect(queryByText(addGroupingText)).not.toBeInTheDocument();
  });

  it('should update grouping', () => {
    clickOnGrouping(detailsColumn.name);

    selectAColumn(createdAtColumn);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it(`should add foreign column grouping properly
      and hide already added column`, () => {
    clickAddMoreGrouping();

    // hide already added column
    const columnSelector = getByTestId('column-selector');
    columns.forEach(column => {
      if (column.name === createdAtColumn.name) {
        expect(
          queryByTextOn(columnSelector, createdAtColumn.name)
        ).not.toBeInTheDocument();
      } else {
        expect(queryByTextOn(columnSelector, column.name)).toBeInTheDocument();
      }
    });

    const productColumn = columns[5];
    const {
      foreignColumns: foreignProductColumns,
      foreignColumnsPanel: foreignProductColumnsPanel
    } = selectAColumn(productColumn, { expand: true });
    const currencyColumn = foreignProductColumns[4];
    const { foreignColumns: foreignCurrencyColumns } = selectAColumn(
      currencyColumn,
      {
        expand: true,
        container: foreignProductColumnsPanel
      }
    );
    const currencyName = foreignCurrencyColumns[1];
    selectAColumn(currencyName);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should remove grouping', () => {
    removeAGrouping();
    removeAGrouping();

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should show the add grouping placeholder again', () => {
    expect(getByText(addGroupingText)).toBeInTheDocument();
  });
});
