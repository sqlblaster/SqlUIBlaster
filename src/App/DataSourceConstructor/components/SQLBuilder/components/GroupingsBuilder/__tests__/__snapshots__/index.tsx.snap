// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Groupings Builder should add a grouping 1`] = `
"SELECT
  \\"orders1\\".\\"Details\\" AS \\"Details (orders1)\\"
FROM
  \\"orders\\" \\"orders1\\"
GROUP BY
  \\"orders1\\".\\"Details\\"
ORDER BY
  \\"orders1\\".\\"Details\\" ASC
LIMIT 2000
"
`;

exports[`Groupings Builder should add foreign column grouping properly
      and hide already added column 1`] = `
"SELECT
  \\"orders1\\".\\"Created At\\" AS \\"Created At (orders1)\\",
  \\"currency1\\".\\"name\\" AS \\"name (currency1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"currency\\" \\"currency1\\" ON \\"products1\\".\\"CurrencyID\\" = \\"currency1\\".\\"ID\\"
GROUP BY
  \\"orders1\\".\\"Created At\\",
  \\"currency1\\".\\"name\\"
ORDER BY
  \\"orders1\\".\\"Created At\\" ASC,
  \\"currency1\\".\\"name\\" ASC
LIMIT 2000
"
`;

exports[`Groupings Builder should remove grouping 1`] = `
"SELECT
  *
FROM
  \\"orders\\"
LIMIT 2000
"
`;

exports[`Groupings Builder should update grouping 1`] = `
"SELECT
  \\"orders1\\".\\"Created At\\" AS \\"Created At (orders1)\\"
FROM
  \\"orders\\" \\"orders1\\"
GROUP BY
  \\"orders1\\".\\"Created At\\"
ORDER BY
  \\"orders1\\".\\"Created At\\" ASC
LIMIT 2000
"
`;
